### ISSUE TITLE
torch.linalg.svd on ROCm: driver= selects nothing but still disables the gesvd fallback

### ISSUE THREAD

**Issue by demandal25** [pytorch/pytorch#198425]
## Summary

On ROCm, `torch.linalg.svd(..., driver=...)` does not select a driver, but it does still disable the gesvd fallback. Passing `driver="gesvd"` — the natural workaround when the Jacobi driver misbehaves — is therefore strictly worse than passing nothing at all.

This surfaced while tracking down a rocSOLVER GESVDJ accuracy bug, where all four driver values produced bit-identical wrong answers on gfx950.

## 1. `driver=` selects nothing on ROCm

`aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.cpp:658-661`:

```cpp
#ifdef USE_ROCM
  const auto driver_v = std::string_view("gesvdj");
#else
  const auto driver_v = driver.value_or("gesvdj");
#endif
```

`driver` is not consulted for dispatch after this, so `driver=None`, `"gesvd"`, `"gesvdj"` and `"gesvda"` all take the gesvdj branch. The `TORCH_CHECK(false, "unknown svd driver")` at `:681` is unreachable on ROCm, so an invalid driver string is silently accepted too.

`torch.linalg.svd` documents `driver` as cuSOLVER-only, so selecting nothing on ROCm is defensible. The next part is not.

## 2. but it still suppresses the fallback

`:694-702`:

```cpp
      if (!driver.has_value()) {
        svd_cusolver_gesvd(A, U, S, V, info, full_matrices, compute_uv, false, svd_non_converging_batches);
      }
```

`driver.has_value()` is read on ROCm even though the value was ignored, so passing any driver removes the gesvd fallback while giving nothing in return. The warning text at `:695-699` also tells ROCm users to "redo this SVD with another driver", which cannot help them.

Suggested fix: on ROCm either ignore `driver` consistently (so the fallback still runs), or warn that it has no effect.

## 3. the fallback was unreachable anyway

The fallback is gated on gesvdj reporting non-convergence. rocSOLVER's small-size Jacobi kernel never set `info = 1` for `n <= 58`, so for every shape PyTorch routes to `gesvdjBatched` (`m, n <= 32`, `:670-675`) it could not fire. That is a rocSOLVER bug, reported as ROCm/rocm-libraries#12501 and fixed in ROCm/rocm-libraries#12502; noting it here because it is why this went unnoticed.

## 4. minor: tolerance is 0 for complex inputs

`:388` and `:464`:

```cpp
TORCH_CUSOLVER_CHECK(cusolverDnXgesvdjSetTolerance(gesvdj_params, std::numeric_limits<scalar_t>::epsilon()));
```

`scalar_t` is the complex type for complex inputs, and neither the standard nor `c10/util/complex.h` specializes `std::numeric_limits` for it, so the primary template returns `T()`, i.e. 0. Verified: `std::numeric_limits<std::complex<float>>::epsilon()` is `0` with `is_specialized == 0`.

The intent is clearly machine epsilon. On ROCm the effect is benign, because rocSOLVER maps `abstol <= 0` to machine precision, so this is a latent wart rather than a live bug there. `std::numeric_limits<value_t>::epsilon()` would express the intent.

## Reproducer

```python
import torch
A = torch.tensor([[1.0, 0.5, 0.0], [0.5, 1.0, 0.0], [0.0, 0.0, 4096.0]]).cuda()
for d in (None, "gesvd", "gesvdj", "gesvda"):
    U, S, Vh = torch.linalg.svd(A, driver=d)
    print(d, S.tolist(), (U @ Vh - torch.eye(3, device="cuda")).abs().max().item())
```

Every driver prints `S = [4096.0, 1.118034, 0.670820]` against an exact `[4096, 1.5, 0.5]`, and `max|UVh - I| = 4.47e-01`. Measured on MI355X (gfx950), torch 2.12.1+rocm7.2.

The wrong values are the rocSOLVER bug; that all four drivers give the same wrong values, and that `driver="gesvd"` makes recovery less likely rather than more, is this issue.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang @malfet @jianyuh @nikitaved @mruberry @walterddr @Lezcano

**Comment by liminfei-amd** [comment]
Reproduced on an RX 9070/gfx1201 with ROCm 10 using the same PyTorch source
SHA before and after the patch.

On ROCm, the documented driver names are currently ignored, but merely
passing one suppresses the existing `gesvd` fallback; invalid names are also
accepted. Empty inputs additionally return before the normal backend and
driver validation.

I prepared a local PyTorch patch that validates backend/name semantics before
the empty-input path and prevents an ignored ROCm driver argument from
suppressing fallback. Focused `svd`/`svdvals`, `out=`, empty-input, CPU, and
ROCm tests pass.

This is separate from ROCm/rocm-libraries#12502, which fixes the underlying
rocSOLVER Jacobi convergence/accuracy issue.
